streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java

```diff
@@ -87,8 +71,13 @@ StampedRecord nextRecord(final RecordInfo info) {
 
         final RecordQueue queue = queuesByTime.poll();
         if (queue != null) {
+            final long oldTimestamp = queue.timestamp();
             // get the first record from this queue.
             record = queue.poll();
+            final long newTimestamp = queue.timestamp();
+            if (newTimestamp != oldTimestamp && newTimestamp > streamTime) {
```

Comment:
Just for my own education: If we change "stream-time" to be computed base on head-records only, can we simplify this to `newTimestamp > streamTime`? Or maybe even simplify the whole `computeStreamTime()`?

Reply:
I think the computation can be:

1. if (newTimestamp != oldTimestamp && oldTimestamp == streamTime) then recompute stream time.
2. in `computeStreamTime()`, I'm not sure if there's much we can further optimize.

Reply:
I'm not sure... We want stream time to be `max(previousStreamTime, [queue times].min)`. I think this definition will remain the same, and we'll just plan to alter the definition of "queue time" (from min enqueued time to head-of-the-queue time).

So the queue-time computation will definitely get simpler.

I suppose that either way, we'll poll from the queue with the smallest queue-time. Which means that before the poll, that queue's timestamp must be `<= streamTime`. So, if, after the poll, that queue's timestamp is `> streamTime`, we may have to advance stream-time.

Now that I think about it, since we're storing the queues in the order of their queue-time, maybe we don't have to iterate over all the queues, but just look at the next one in the priority queue. It alone should tell us what the new value of `streamTime` should be...

I'll try this out tomorrow.